File: cs/algo/search/linear_search.rb
# frozen_string_literal: true

# Linear Search in Ruby (index version)
#
# Returns the index of the first element that matches the target.
# Returns nil if the target is not found.
# This is equivalent to Ruby's Array#find_index.
#
# Time Complexity: O(N)
def linear_search(array, target)
  array.each_with_index do |value, index|
    return index if value == target
  end
  nil
end

# --- Demo ---
# puts linear_search([10, 20, 30, 40, 50], 30) # Output: 2

# Linear Search in Ruby (value version)
#
# Returns the *value* of the first element that matches the target.
# Returns nil if the target is not found.
# This is equivalent to Ruby's Array#find.
#
# Time Complexity: O(N)
def linear_search_value(array, target)
  array.each do |value|
    return value if value == target
  end
  nil
end

# --- Demo ---
puts linear_search_value([10, 20, 30, 40, 50], 30) # Output: 30

File: cs/algo/search/linear_search.go
package main

import (
	"fmt"
)

// LinearSearch in Go (index version)
//
// Returns the index of the first element that matches the target.
// Returns -1 if the target is not found.
//
// Time Complexity: O(N)
func LinearSearch(arr []int, target int) int {
	for i, v := range arr {
		if v == target {
			return i
		}
	}
	return -1 // index does not become -1, so we use -1 to indicate "not found"
}

// LinearSearchValue in Go (value version)
//
// Returns the *value* of the first element that matches the target.
// Returns -1 if the target is not found.
//
// Time Complexity: O(N)
func LinearSearchValue(arr []int, target int) (int, bool) {
	for _, v := range arr {
		if v == target {
			return v, true
		}
	}
	return -1, false // value possibly -1, so we return a bool to indicate "not found"
}

func main() {
	arr := []int{10, 20, 30, 40, 50}

	// --- Demo ---
	fmt.Println(LinearSearch(arr, 30))      // Output: 2
	fmt.Println(LinearSearchValue(arr, 30)) // Output: 30
}
